### PR TITLE
Plane: add AirMode RC option for quadplanes

### PR DESCRIPTION
**Pull request by kd0aij**
add Q_OPTION for AirMode (auto-enabled if RCx_OPTION ARMDISARM assigned). There is no behavior change unless an RC option is assigned for either ARMDISARM or AIRMODE.

This improves control in VTOL modes with the throttle near zero (SPIN_MIN) by preventing throttle suppression and by setting throttle mix to manual when in manual throttle modes.

includes a bugfix for manual throttle mix for qacro

### CODE REVIEW

File: ArduPlane/Plane.h [hunk 1]
```diff
@@ -785,6 +785,9 @@ class Plane : public AP_Vehicle {
     // soaring mode-change timer
     uint32_t soaring_mode_timer;
 
+    // air mode is 0 = not-configured; 1 = disabled; 2 = enabled
```
Comment (andyp1per) [inline ArduPlane/Plane.h:788]: I think this comment is wrong based on the implementation?

**Comment by IamPete1** [comment]
Do we need a Qoption and a switch?

**Comment by Hwurzburg** [comment]
I think Mark was concerned about altering existing operation....if its mirroring Copter it would only need to make airmode active if armed by switch function 41 and then add the air mode disable switch  like copter just added...but that is different than now with switch 41

File: ArduPlane/Plane.h (hunk 1 above)
Comment (kd0aij) [reply ArduPlane/Plane.h:788]: right, I changed from DISABLED/NONE/ENABLED to off/on due to collisions with macros and forgot to update this

**Comment by kd0aij** [comment]
@IamPete1 Without the new q_option (defaulting to 0) this would change behavior for quadplane users who used a switch for arming. 
@Hwurzburg I just realized that this implementation for Q_OPTION=AIRMODE completely overrides the RC_OPTION for airmode; if you set that option bit airmode is always ON when armed, regardless of the RC switch position. But if you use rudder arming, you can turn airmode on and off (while armed) using the assigned switch.

**Comment by Hwurzburg** [comment]
yep,I would think that the option would turn on AIRMODE with no AIRMODE switch (better use an arm/disarm switch then), but that the  AIRMODE switch, if configured, would determine when it was on or off....and disarm/arm switch is independent of all of this....assuming you want to keep the same profile as now...wont be like copter, but ....not a problem I guess...

**Comment by IamPete1** [comment]
I remove the ability for Airmode via arming/disarming switch, that was really accident on copter. If you going to have to have a switch you might as well have it dedicate to Airmode.

**Comment by kd0aij** [comment]
@Hwurzburg Tested in SITL:
If you set the new Q_OPTIONS bit 7, and have arming on a switch, you'll have airmode at all times when armed. 

If rudder arming is enabled, then you can't get airmode unless you set an RC option to 84.  Then the switch controls airmode, regardless of rudder_arming_type and Q_OPTIONS.

I believe this now behaves the same as copter only if the Q_OPTIONS bit is set, and does not change existing behavior unless the new RC_OPTION is assigned.

**Comment by Hwurzburg** [comment]
Mark,  with option bit set...is it when rudder arming is enabled? or used to arm? that the switch is needed to get into AIRMODE....in copter if you arm by rudder, you need the switch....so the option bit is only for airmode on arm switch, right? with no arm/disarm you need the airmode switch, right?

also, I think you need to update RC_Channel.cpp with the AIRMODE switch metadata now that Plane uses it,right?

**Comment by kd0aij** [comment]
I forgot there was separate metadata for each vehicle... will update the plane entry.

Is my description of the options bit and rc_option above ambiguous or unclear?  That is essentially the same text I'd propose to add to the wiki.

**Comment by Hwurzburg** [comment]
"If you set the new Q_OPTIONS bit 7, and have arming on a switch, you'll have airmode at all times when armed." okay and what if you armed with the rudder?immediate air mode and jumping around until you release rudder? is AIRMODE active then? this is not what Copter does, so I do think it might be unclear a bit...thats why I asked?

**Comment by kd0aij** [comment]
The next sentence is "If rudder arming is enabled, then you can't get airmode unless you set an RC option to AIRMODE. Then the switch controls airmode, regardless of rudder_arming_type and Q_OPTIONS."

I can't see how that doesn't answer your question, but maybe I'm missing something?

**Comment by Hwurzburg** [comment]
Its just a matter of having every statement true on its own and not qualified by later statements. I would word it thus:

If the Q_OPTIONS bit 7  is set to "1", and only have arming possible via an ARM/DISARM RCx_OPTIONS switch, or via GCS arm command, then AIRMODE will be active at all times when armed.
If rudder arming is enabled (ARMING_RUDDER = 1 or 2), then AIRMODE can only be activated by using an AIRMODE RCx_OPTIONS switch.

**Comment by kd0aij** [comment]
OK, I thought you were saying the behavior was not what you wanted. 
If this gets merged, I'll let you write the wiki entry :)

https://github.com/IamPete1 In reply to your last comment the Q_OPTIONS bit for airmode on arming was specifically requested by Henry.

**Comment by Hwurzburg** [comment]
so that people migrating from Copter to QuadPlane have the same experience in VTOL modes

File: ArduPlane/quadplane.cpp [hunk 2]
```diff
@@ -1896,7 +1908,7 @@ void QuadPlane::update_throttle_suppression(void)
     }
 
     // if the users throttle is above zero then allow motors to run
-    if (plane.get_throttle_input() != 0) {
+    if ((plane.get_throttle_input() != 0) || (plane.air_mode == AirMode::ON)) {
```
Comment (tridge) [inline ArduPlane/quadplane.cpp:1911]: should this be on all flight modes?

**Comment by kd0aij** [comment]
check armingmethod instead of rudder arming
use accessor to qualify throttle suppression

**Comment by kd0aij** [comment]
@Hwurzburg I changed this to limit auto-airmode to switch arming only:
If you set the new Q_OPTIONS bit 7, and arm using a switch, arming/disarming also turns airmode on/off.
If you assign an aux switch for AirMode, that switch controls airmode, with no exceptions.

**Comment by Hwurzburg** [comment]
super! sorry my last addition causes you to fix and rebase!

File: ArduPlane/quadplane.cpp [hunk 3]
```diff
@@ -1398,6 +1398,18 @@ void QuadPlane::set_armed(bool armed)
     if (!initialised) {
         return;
     }
+    // If OPTION_AIRMODE is selected and no AIRMODE switch is assigned and
+    // rudder arming is disabled (don't want vehicle to yaw when disarming with rudder)
+    if ((options & OPTION_AIRMODE) && 
+        (rc().find_channel_for_option(RC_Channel::AUX_FUNC::AIRMODE) == nullptr) &&
+        (plane.arming.get_rudder_arming_type() == AP_Arming::RudderArming::IS_DISABLED)) {
+        // automatically enable/disable AirMode on arming/disarming
```
[the remark this answers is not in the document]
Comment (kd0aij) [reply ArduPlane/quadplane.cpp]: addressed this by moving the logic into the arm/disarm functions

File: ArduPlane/quadplane.cpp (hunk 2 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp]: changed to apply only to non-auto throttle modes

File: ArduPlane/quadplane.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (kd0aij) [reply ArduPlane/quadplane.cpp]: reworked the logic to apply only to switch arming

File: Tools/autotest/quadplane.py [hunk 4]
```diff
@@ -78,6 +78,144 @@ def get_disarm_delay(self):
...
+        self.disarm_vehicle()
+
+        """arm with switch, record Motor1 SPIN_MIN PWM output and disarm"""
+        self.set_rc(8, 2000)
+        self.progress("Waiting for Motor1 to spool up to SPIN_MIN")
+        spin_min_pwm = self.get_servo_pwm(5, min_pwm)
+        self.progress("spin_min_pwm: %d" % spin_min_pwm)
+        self.set_rc(8, 1000)
+
+        assert(spin_arm_pwm < spin_min_pwm)
```
Comment (peterbarker) [inline Tools/autotest/quadplane.py:133]: This assert could be for the values your checking against rather than received; it's a sensible precondition.

File: Tools/autotest/quadplane.py [hunk 5]
```diff
@@ -78,6 +78,144 @@ def get_disarm_delay(self):
...
+
+        """verify that arming with switch results in higher PWM output"""
+        self.progress("Test auxswitch arming with Q_OPTIONS=AirMode")
+        modes = ('QSTABILIZE', 'QACRO')
```
Comment (peterbarker) [inline Tools/autotest/quadplane.py:137]: Maybe just inline this.

File: Tools/autotest/quadplane.py [hunk 6]
```diff
@@ -78,6 +78,144 @@ def get_disarm_delay(self):
...
+        self.change_mode('QSTABILIZE')
+        self.zero_throttle()
+        self.wait_ready_to_arm()
+
+        """SPIN_ARM and SPIN_MIN default to 0.10 and 0.15"""
+        """when armed with zero throttle in AirMode, motor PWM should be at SPIN_MIN"""
+        """If AirMode is off, motor PWM will drop to SPIN_ARM"""
+
+        """Make sure that SERVO5 is Motor1 (default)"""
+        assert(33 == self.get_parameter('SERVO5_FUNCTION'))
```
[the remark this answers is not in the document]
Comment (kd0aij) [reply Tools/autotest/quadplane.py:103]: I guess there's something unusual about the eye of this beholder :)
will look at PreconditionFailed

File: Tools/autotest/quadplane.py [hunk 7]
```diff
@@ -78,6 +78,144 @@ def get_disarm_delay(self):
...
+
+        """Make sure that flightmode channel is 5 (default)"""
+        assert(5 == self.get_parameter("FLTMODE_CH"))
+
+        """When disarmed, motor PWM will drop to min_pwm"""
+        #min_pwm = self.get_parameter('SERVO5_MIN')
+        """SERVO5_MIN is not respected; it defaults to 1100, but disarmed PWM is 1000"""
```
[the remark this answers is not in the document]
Comment (kd0aij) [reply Tools/autotest/quadplane.py:110]: Thanks, didn't realize that

File: Tools/autotest/quadplane.py [hunk 8]
```diff
@@ -78,6 +78,144 @@ def get_disarm_delay(self):
...
+        """set Q_OPTIONS bit AIRMODE"""
+        self.set_parameter("Q_OPTIONS", 512)
+
+        """set  RC8_OPTION to ARMDISARM"""
+        self.set_parameter("RC8_OPTION", 41)
+
+        """arm with GCS, record Motor1 SPIN_ARM PWM output and disarm"""
+        self.arm_vehicle()
+        self.progress("Waiting for Motor1 to spool up to SPIN_ARM")
+        spin_arm_pwm = self.get_servo_pwm(5, min_pwm)
```
[the remark this answers is not in the document]
Comment (kd0aij) [reply Tools/autotest/quadplane.py:122]: just changed wait_servo_channel_value to return the desired value...
I hope that's reasonably pythonic

File: Tools/autotest/quadplane.py (hunk 4 above)
Comment (kd0aij) [reply Tools/autotest/quadplane.py:133]: good point

File: Tools/autotest/quadplane.py (hunk 5 above)
Comment (kd0aij) [reply Tools/autotest/quadplane.py:137]: wouldn't that add a dozen lines of code?

File: Tools/autotest/quadplane.py [hunk 9]
```diff
@@ -78,6 +78,144 @@ def get_disarm_delay(self):
     def set_autodisarm_delay(self, delay):
         self.set_parameter("LAND_DISARMDELAY", delay)
 
+    def get_servo_pwm(self, channel, min_pwm):
+        self.delay_sim_time(2)
+        self.wait_servo_channel_value(channel, min_pwm, comparator=operator.gt)
+        m = self.mav.recv_match(type='SERVO_OUTPUT_RAW', blocking=True, timeout=1)
+        if m is None:
+            raise NotAchievedException("Did not get SERVO_OUTPUT_RAW")
+        pwm = getattr(m, "servo5_raw", None)
```
[the remark this answers is not in the document]
Comment (kd0aij) [reply Tools/autotest/quadplane.py:87]: function is gone now

File: Tools/autotest/quadplane.py [hunk 10]
```diff
@@ -78,6 +78,144 @@ def get_disarm_delay(self):
     def set_autodisarm_delay(self, delay):
         self.set_parameter("LAND_DISARMDELAY", delay)
 
+    def get_servo_pwm(self, channel, min_pwm):
+        self.delay_sim_time(2)
```
[the remark this answers is not in the document]
Comment (kd0aij) [reply Tools/autotest/quadplane.py:82]: spool-up delay... moved to timeout

File: Tools/autotest/quadplane.py [hunk 11]
```diff
@@ -78,6 +78,144 @@ def get_disarm_delay(self):
...
+            raise NotAchievedException("Did not get SERVO_OUTPUT_RAW")
+        pwm = getattr(m, "servo5_raw", None)
+        return pwm
+
+    def test_airmode(self):
+        """Check that plane.air_mode turns on and off as required"""
+        self.progress("########## Testing AirMode operation")
+        self.set_parameter("AHRS_EKF_TYPE", 10)
+        self.change_mode('QSTABILIZE')
+        self.zero_throttle()
```
[the remark this answers is not in the document]
Comment (kd0aij) [reply Tools/autotest/quadplane.py:95]: removed

File: Tools/autotest/quadplane.py [hunk 12]
```diff
@@ -78,6 +78,144 @@ def get_disarm_delay(self):
...
+        """Make sure that flightmode channel is 5 (default)"""
+        assert(5 == self.get_parameter("FLTMODE_CH"))
+
+        """When disarmed, motor PWM will drop to min_pwm"""
+        #min_pwm = self.get_parameter('SERVO5_MIN')
+        """SERVO5_MIN is not respected; it defaults to 1100, but disarmed PWM is 1000"""
+        min_pwm = 1000
+
+        """set Q_OPTIONS bit AIRMODE"""
+        self.set_parameter("Q_OPTIONS", 512)
```
[the remark this answers is not in the document]
Comment (kd0aij) [reply Tools/autotest/quadplane.py:114]: since I haven't considered whether any of the other options might affect this test,
I'll force them all off to make it somewhat apparent.

File: ArduPlane/AP_Arming.cpp [hunk 13]
```diff
@@ -179,6 +179,14 @@ bool AP_Arming_Plane::arm(const AP_Arming::Method method, const bool do_arming_c
         return false;
     }
 
+    if ((method == Method::AUXSWITCH) && (plane.quadplane.options & QuadPlane::OPTION_AIRMODE)) {
+        // if no airmode switch assigned
```
Comment (peterbarker) [inline ArduPlane/AP_Arming.cpp:183]:
```suggestion
        // if no airmode switch assigned then we honour the QuadPlane option bit:
```

File: ArduPlane/AP_Arming.cpp [hunk 14]
```diff
@@ -179,6 +179,14 @@ bool AP_Arming_Plane::arm(const AP_Arming::Method method, const bool do_arming_c
...
+        RC_Channel *c = rc().find_channel_for_option(RC_Channel::AUX_FUNC::AIRMODE);
```
Comment (peterbarker) [inline ArduPlane/AP_Arming.cpp:184]: Can do without this temp

File: ArduPlane/AP_Arming.cpp [hunk 15]
```diff
@@ -202,6 +210,12 @@ bool AP_Arming_Plane::disarm(const AP_Arming::Method method)
     // suppress the throttle in auto-throttle modes
     plane.throttle_suppressed = plane.auto_throttle_mode;
 
+    // if no airmode switch assigned, ensure airmode is off
```
Comment (peterbarker) [inline ArduPlane/AP_Arming.cpp:213]:
```suggestion
    // if no airmode switch assigned, ensure airmode is off:
```

File: ArduPlane/AP_Arming.cpp [hunk 16]
```diff
@@ -202,6 +210,12 @@ bool AP_Arming_Plane::disarm(const AP_Arming::Method method)
...
+    RC_Channel *c = rc().find_channel_for_option(RC_Channel::AUX_FUNC::AIRMODE);
```
Comment (peterbarker) [inline ArduPlane/AP_Arming.cpp:214]: can do without this temp

File: ArduPlane/Plane.h [hunk 17]
```diff
@@ -789,6 +789,9 @@ class Plane : public AP_Vehicle {
     bool non_auto_terrain_disable;
     bool terrain_disabled();
 
+    // air mode state: off=0, on=1
```
Comment (peterbarker) [inline ArduPlane/Plane.h:792]: This is now an enumeration; shouldn't be hinting at numbers here.

File: ArduPlane/quadplane.cpp [hunk 18]
```diff
@@ -3284,9 +3289,9 @@ void QuadPlane::update_throttle_mix(void)
         return;
     }
 
-    if (plane.control_mode == &plane.mode_qstabilize) {
+    if ((plane.control_mode == &plane.mode_qstabilize) || (plane.control_mode == &plane.mode_qhover) || (plane.control_mode == &plane.mode_qacro)) {
```
Comment (peterbarker) [inline ArduPlane/quadplane.cpp:3292]: This change doesn't seem related to the PR title and would appear to be a significant behavioural change?

File: Tools/autotest/quadplane.py [hunk 19]
```diff
@@ -78,6 +78,147 @@ def get_disarm_delay(self):
     def set_autodisarm_delay(self, delay):
         self.set_parameter("LAND_DISARMDELAY", delay)
 
+    def test_airmode(self):
+        """Check that plane.air_mode turns on and off as required"""
+        self.progress("########## Testing AirMode operation")
+        self.set_parameter("AHRS_EKF_TYPE", 10)
```
Comment (peterbarker) [inline Tools/autotest/quadplane.py:84]: It's no biggy - but why change EKF types?

File: Tools/autotest/quadplane.py [hunk 20]
```diff
@@ -78,6 +78,147 @@ def get_disarm_delay(self):
...
+        self.change_mode('QSTABILIZE')
+        self.wait_ready_to_arm()
+
+        """SPIN_ARM and SPIN_MIN default to 0.10 and 0.15"""
+        """when armed with zero throttle in AirMode, motor PWM should be at SPIN_MIN"""
+        """If AirMode is off, motor PWM will drop to SPIN_ARM"""
```
Comment (peterbarker) [inline Tools/autotest/quadplane.py:90]:
```suggestion
        """
        SPIN_ARM and SPIN_MIN default to 0.10 and 0.15
        when armed with zero throttle in AirMode, motor PWM should be at SPIN_MIN
        If AirMode is off, motor PWM will drop to SPIN_ARM
        """
```

`"""` start multi-line comments :-)

Indenting may need fixing on my suggestion there.

File: Tools/autotest/quadplane.py [hunk 21]
```diff
@@ -78,6 +78,147 @@ def get_disarm_delay(self):
...
+
+        """Make sure that SERVO5 is Motor1 (default)"""
```
Comment (peterbarker) [inline Tools/autotest/quadplane.py:92]:
```suggestion
        self.progress("Make sure that SERVO5 is Motor1 (default)")
```

File: Tools/autotest/quadplane.py [hunk 22]
```diff
@@ -78,6 +78,147 @@ def get_disarm_delay(self):
...
+        Motor1_value = 33
```
Comment (peterbarker) [inline Tools/autotest/quadplane.py:93]: allow-lower-case for variable names.

This one should probably be caused `motor1_servo_function_number`

My initial reaction was this was an RPM.  And I prefer long play.

File: Tools/autotest/quadplane.py [hunk 23]
```diff
@@ -78,6 +78,147 @@ def get_disarm_delay(self):
...
+        if (self.get_parameter('SERVO5_FUNCTION') != Motor1_value):
+            raise PreconditionFailedException("SERVO5_FUNCTION not %d" % Motor1_value)
+
+        """Make sure that flightmode channel is 5 (default)"""
```
Comment (peterbarker) [inline Tools/autotest/quadplane.py:97]:
```suggestion
        self.progress("Make sure that flightmode channel is 5 (default)")
```

Similarly for other useful comments.

File: Tools/autotest/quadplane.py [hunk 24]
```diff
@@ -78,6 +78,147 @@ def get_disarm_delay(self):
...
+        default_fltmode_ch = 5
```
Comment (peterbarker) [inline Tools/autotest/quadplane.py:98]: if this fails then the harness has failed.  I don't mind it being here, 'though.

File: Tools/autotest/quadplane.py [hunk 25]
```diff
@@ -78,6 +78,147 @@ def get_disarm_delay(self):
...
+        if (self.get_parameter("FLTMODE_CH") != default_fltmode_ch):
+            raise PreconditionFailedException("FLTMODE_CH not %d" % default_fltmode_ch)
+
+        """When disarmed, motor PWM will drop to min_pwm"""
+        #min_pwm = self.get_parameter('SERVO5_MIN')
+        """SERVO5_MIN is not respected; it defaults to 1100, but disarmed PWM is 1000"""
```
Comment (peterbarker) [inline Tools/autotest/quadplane.py:104]: Comment needs adjustment.

Actually fiddling with `Q_THR_MIN_PWM` a little to make sure it is honoured might be nice.

File: Tools/autotest/quadplane.py [hunk 26]
```diff
@@ -78,6 +78,147 @@ def get_disarm_delay(self):
...
+        armdisarm_option = 41
+        arm_ch = 8
+        self.set_parameter("RC%d_OPTION" % arm_ch, armdisarm_option)
+
+        """arm with GCS, record Motor1 SPIN_ARM PWM output and disarm"""
+        spool_delay = self.get_parameter("Q_M_SPOOL_TIME") + 0.25
+        self.zero_throttle()
+        self.arm_vehicle()
+        self.progress("Waiting for Motor1 to spool up to SPIN_ARM")
+        self.delay_sim_time(spool_delay)
```
Comment (peterbarker) [inline Tools/autotest/quadplane.py:121]: Don't ned this if you're just going to wait on the servo channel value; you can increase the timeout if you need to.

Similarly elsewhere.

File: Tools/autotest/quadplane.py [hunk 27]
```diff
@@ -78,6 +78,147 @@ def get_disarm_delay(self):
...
+        self.progress("spin_min_pwm: %d" % spin_min_pwm)
+        self.set_rc(8, 1000)
+
+        if (spin_arm_pwm >= spin_min_pwm):
+            raise PreconditionFailedException("SPIN_MIN pwm not greater than SPIN_ARM pwm")
+
+        """verify that arming with switch results in higher PWM output"""
+        self.progress("########## Test auxswitch arming with Q_OPTIONS=AirMode")
+        modes = ('QSTABILIZE', 'QACRO')
+        for mode in modes:
```
Comment (peterbarker) [inline Tools/autotest/quadplane.py:140]:
```suggestion
        for mode in ['QSTABILIZE', 'QACRO']:
```

File: Tools/autotest/quadplane.py (hunk 5 above)
Comment (peterbarker) [reply Tools/autotest/quadplane.py:139]: Suggestion added.

Likewise for the loop below this one.

File: Tools/autotest/quadplane.py [hunk 28]
```diff
@@ -78,6 +78,147 @@ def get_disarm_delay(self):
...
+            self.progress("arming with switch at zero throttle")
+            self.arm_motors_with_switch(arm_ch)
+            self.progress("Waiting for Motor1 to speed up")
+            self.wait_servo_channel_value(5, spin_min_pwm, comparator=operator.ge)
+
+            self.progress("disarming with switch")
+            self.disarm_motors_with_switch(arm_ch)
+            self.progress("Waiting for Motor1 to stop")
+            self.wait_servo_channel_value(5, min_pwm, comparator=operator.le)
+            self.wait_ready_to_arm()
```
Comment (peterbarker) [inline Tools/autotest/quadplane.py:153]: We usually only do this going into a test rather than leaving it.  Not fussed 'though.

File: Tools/autotest/quadplane.py [hunk 29]
```diff
@@ -78,6 +78,147 @@ def get_disarm_delay(self):
...
+            self.wait_servo_channel_value(5, spin_arm_pwm, comparator=operator.le)
+
+            self.progress("disarming with switch")
+            self.disarm_motors_with_switch(arm_ch)
+            self.progress("Waiting for Motor1 to stop")
+            self.wait_servo_channel_value(5, min_pwm, comparator=operator.le)
+            self.wait_ready_to_arm()
+
+        """set  RC7_OPTION to AIRMODE"""
+        """verify that RC7 turns airmode on/off while armed"""
```
Comment (peterbarker) [inline Tools/autotest/quadplane.py:174]: This is somewhere `start_subtest` might be good.
```suggestion
        self.start_subtest("verify that RC7 turns airmode on/off while armed")
```

File: ArduPlane/AP_Arming.cpp (hunk 14 above)
Comment (kd0aij) [reply ArduPlane/AP_Arming.cpp:184]: done

File: ArduPlane/AP_Arming.cpp (hunk 15 above)
Comment (kd0aij) [reply ArduPlane/AP_Arming.cpp:213]: done

File: ArduPlane/AP_Arming.cpp (hunk 16 above)
Comment (kd0aij) [reply ArduPlane/AP_Arming.cpp:214]: done

File: ArduPlane/Plane.h (hunk 17 above)
Comment (kd0aij) [reply ArduPlane/Plane.h:792]: done

File: Tools/autotest/quadplane.py (hunk 27 above)
Comment (kd0aij) [reply Tools/autotest/quadplane.py:140]: done

File: Tools/autotest/quadplane.py (hunk 28 above)
Comment (kd0aij) [reply Tools/autotest/quadplane.py:153]: noted

File: Tools/autotest/quadplane.py (hunk 29 above)
Comment (kd0aij) [reply Tools/autotest/quadplane.py:174]: done

File: ArduPlane/AP_Arming.cpp (hunk 13 above)
Comment (kd0aij) [reply ArduPlane/AP_Arming.cpp:183]: done

File: Tools/autotest/quadplane.py (hunk 26 above)
Comment (kd0aij) [reply Tools/autotest/quadplane.py:121]: have to wait long enough for the motor to spool all the way up here, or else spin_arm_pwm is incorrect

**Comment by kd0aij** [comment]
@peterbarker Thanks again. I think this PR is about ready to go now.
I lost track of your comment on throttle mixing; that *is* related to airmode; needs to be manual if airmode is on

File: ArduPlane/quadplane.cpp (hunk 18 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp:3292]: this is the QACRO bugfix; related because airmode still needs manual throttle mix when throttle is zero
I don't think this applies to qhover mode though; will remove it from the test.

File: Tools/autotest/quadplane.py (hunk 19 above)
Comment (kd0aij) [reply Tools/autotest/quadplane.py:84]: EKF is irrelevant, so no point in simulating it here

File: Tools/autotest/quadplane.py (hunk 20 above)
Comment (kd0aij) [reply Tools/autotest/quadplane.py:90]: done

File: Tools/autotest/quadplane.py (hunk 22 above)
Comment (kd0aij) [reply Tools/autotest/quadplane.py:95]: :) renamed

File: Tools/autotest/quadplane.py (hunk 25 above)
Comment (kd0aij) [reply Tools/autotest/quadplane.py]: done

File: Tools/autotest/quadplane.py [hunk 30]
```diff
@@ -78,6 +78,146 @@ def get_disarm_delay(self):
...
+        if (self.get_parameter("FLTMODE_CH") != default_fltmode_ch):
+            raise PreconditionFailedException("FLTMODE_CH not %d" % default_fltmode_ch)
+
+        """When disarmed, motor PWM will drop to min_pwm"""
+        min_pwm = self.get_parameter("Q_THR_MIN_PWM")
+
+        self.progress("Verify Motor1 is ast min_pwm when disarmed")
+        self.wait_servo_channel_value(5, min_pwm, comparator=operator.eq)
+
+        """set Q_OPTIONS bit AIRMODE"""
```
Comment (peterbarker) [inline Tools/autotest/quadplane.py:110]: Just a note that '# ' comments still work :-)  No action required.

**Comment by andyp1per** [comment]
I probably should port the airmode test to copter as well

File: Tools/autotest/quadplane.py (hunk 30 above)
Comment (kd0aij) [reply Tools/autotest/quadplane.py:110]: I just needed to change the color of # comments in my editor to be able to see them
Eyes are getting older and weaker :(

**Comment by kd0aij** [comment]
I've been playing with pulling the parameter values from the metadata here:
https://github.com/kd0aij/ardupilot/pull/22

File: ArduPlane/quadplane.cpp [hunk 31]
```diff
@@ -1920,6 +1920,11 @@ void QuadPlane::update_throttle_suppression(void)
         return;
     }
 
+    // if not in an auto throttle mode and air_mode is on then allow motors to run
+    if (!plane.auto_throttle_mode && air_mode == AirMode::ON) {
```
[the remark this answers is not in the document]
Comment (kd0aij) [reply ArduPlane/quadplane.cpp]: added method Mode::is_vtol_man_throttle() to constrain this to qstabilize and qacro modes

sorry it took two iterations to get this right, but I think that code inspection is now sufficient to rule out behavior changes in other modes.

File: ArduPlane/quadplane.cpp [hunk 32]
```diff
@@ -1920,6 +1920,11 @@ void QuadPlane::update_throttle_suppression(void)
         return;
     }
 
+    // if in a VTOL manual throttle mode and air_mode is on then allow motors to run
+    if (plane.control_mode->is_vtol_man_throttle() && air_mode == AirMode::ON) {
```
Comment (kd0aij) [inline ArduPlane/quadplane.cpp:1924]: fixed using new accessor in Mode class

**Comment by kd0aij** [comment]
QSTABILIZE mode: takeoff then cut throttle to zero, first with airmode off, then with airmode on
https://youtu.be/DFXEZXBejnE

**Comment by tridge** [comment]
Henry will do testing

**Comment by kd0aij** [comment]
@peterbarker I added a commit to extend the autotest with the additional testing that Tridge suggested. Does that change look OK to you?

File: Tools/autotest/quadplane.py [hunk 33]
```diff
@@ -164,12 +168,17 @@ def test_airmode(self):
...
+            self.wait_servo_channel_value(7, spin_arm_pwm, comparator=operator.le)
+            self.wait_servo_channel_value(8, spin_arm_pwm, comparator=operator.le)
 
             self.progress("disarming with switch")
             self.disarm_motors_with_switch(arm_ch)
             self.progress("Waiting for Motor1 to stop")
             self.wait_servo_channel_value(5, min_pwm, comparator=operator.le)
             self.wait_ready_to_arm()
+        # remove attitude error
+        self.set_parameter("AHRS_TRIM_X", 60*math.pi/180)
```
Comment (peterbarker) [inline Tools/autotest/quadplane.py:181]: Probably better to revert this to the previous value - so take a backup and restore it here.  You could use a context instead - but probably overkill.

**Comment by peterbarker** [comment]
@kd0aij thanks for fixing that test up.  Just waiting for test results now, I think.

**Comment by tridge** [comment]
thanks to @Hwurzburg for testing!
